core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala

```diff
@@ -20,94 +20,93 @@ import java.nio.ByteBuffer
-
-  // scalastyle:off
-  def this(ds: Dataset[Row], queryString: String,
-      cachedRDD: RDD[InternalRow], shuffleDependencies: Array[Int],
-      rddId: Int, hasLocalCollectProcessing: Boolean,
-      allLiterals: Array[LiteralValue],
-      queryHints: java.util.Map[String, String],
-      planProcessingTime: Long, currentExecutionId: Option[Long]) = {
-    // scalastyle:on
-    this(ds.sparkSession, ds.queryExecution, ds.exprEnc, queryString, cachedRDD,
-      shuffleDependencies, rddId, hasLocalCollectProcessing, allLiterals,
-      queryHints, planProcessingTime, currentExecutionId)
-  }
+class CachedDataFrame(snappySession: SnappySession, queryExecution: QueryExecution,
+    private[sql] val queryExecutionString: String,
+    private[sql] val queryPlanInfo: SparkPlanInfo,
+    private[sql] var currentQueryExecutionString: String,
+    private[sql] var currentQueryPlanInfo: SparkPlanInfo,
+    cachedRDD: RDD[InternalRow], shuffleDependencies: Vector[Int], encoder: Encoder[Row],
+    val rddId: Int, noSideEffects: Boolean, val queryHints: java.util.Map[String, String],
+    private[sql] var currentExecutionId: Long,
+    private[sql] var planStartTime: Long, private[sql] var planEndTime: Long)
+    extends Dataset[Row](snappySession, queryExecution, encoder) with Logging {
 
-  private var lastExecRDD: RDD[InternalRow] = cachedRDD
+  private[sql] final def isCached: Boolean = cachedRDD ne null
 
-  private def getExecRDD: RDD[InternalRow] = if (cachedRDD ne null) cachedRDD else {
-    if (lastExecRDD ne null) lastExecRDD
+  private def getExecRDD: RDD[InternalRow] =
+    if (isCached) cachedRDD else queryExecution.executedPlan.execute()
+
+  @transient
+  private var _boundEnc: ExpressionEncoder[Row] = _
+
+  private def boundEnc: ExpressionEncoder[Row] = {
+    if (_boundEnc ne null) _boundEnc
     else {
-      val rdd = queryExecution.executedPlan.execute()
-      lastExecRDD = rdd
-      rdd
+      _boundEnc = exprEnc.resolveAndBind(logicalPlan.output,
+        snappySession.sessionState.analyzer)
+      _boundEnc
     }
   }
 
-  /**
-   * Return true if [[collectWithHandler]] supports partition-wise separate
-   * result handling by default, else result handler is invoked for a
-   * single partition result.
-   */
-  def hasPartitionWiseHandling: Boolean = cachedRDD ne null
+  private def isLowLatencyQuery: Boolean =
+    isCached && cachedRDD.getNumPartitions <= 2 /* some small number */
 
-  private lazy val boundEnc = exprEnc.resolveAndBind(logicalPlan.output,
-    sparkSession.sessionState.analyzer)
+  @transient
+  private lazy val lastShuffleCleanups = new Array[Future[Unit]](shuffleDependencies.length)
 
-  private lazy val queryExecutionString: String = queryExecution.toString()
+  @transient
+  private var _rowConverter: UnsafeProjection = _
 
-  private lazy val isLowLatencyQuery: Boolean =
-    getExecRDD.getNumPartitions <= 2 /* some small number */
+  private def rowConverter: UnsafeProjection = {
```

Comment:
Have changed to "def" because partition pruning can potentially change so its no longer evaluated once but on every execution. Though after the "duplicate()" addition where original CachedDataFrame is not returned rather a shallow copy, this will not matter.

Reply:
Oh, commented about wrong one.

The reason for doing this way is while duplicating, it can copy over the existing one to the duplicated CDFs rather than creating afresh (which is a bit expensive for point queries in profiling).